### PR TITLE
Pla 1490 virtual device minting

### PR DESCRIPTION
**Pull request by 0xdev22**
# Proposed Changes

### CODE REVIEW

File: internal/controllers/virtual_devices_controller.go [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+// SignVirtualDeviceMintingPayload godoc
+// @Description validate signed signature for vehicle minting
+// @Tags        integrations
+// @Produce     json
+// @Success     200 {array}
+// @Router      /integration/:tokenID/mint-virtual-device [post]
+func (vc *VirtualDeviceController) SignVirtualDeviceMintingPayload(c *fiber.Ctx) error {
+	tokenID := c.Params("tokenID")
+	vehicleNode := c.Query("vehicle_id")
```
Comment (JamesReate) [inline internal/controllers/virtual_devices_controller.go:148]: why would `vehicle_id` not be in the path and instead in the querystring - is it optional?

File: internal/test/helpers.go [hunk 2]
```diff
@@ -515,3 +518,37 @@ func BuildDeviceDefinitionGRPC(deviceDefinitionID string, mk string, model strin
 
 	return []*ddgrpc.GetDeviceDefinitionItemResponse{rp}
 }
+
+func BuildGetUserGRPC(id string, email *string, ethereum_address *string, referred_by *users.UserReferrer) *pb.User {
```
[the remark this answers is not in the document]
Comment (elffjs) [reply internal/test/helpers.go:522]: Similarly, `referredBy`.

File: internal/controllers/virtual_devices_controller.go [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+	uTokenID, err := strconv.ParseUint(tokenID, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid tokenID provided")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
+		Id: userID,
+	})
+	if err != nil {
+		vc.log.Debug().AnErr("GetUser Grpc", err).Msg("error occurred when fetching user")
```
[the remark this answers is not in the document]
Comment (elffjs) [reply internal/controllers/virtual_devices_controller.go:104]: Let's use the gRPC error translator. There is a difference between "can't find you" and "I can't reach the upstream service". It's not super precise but it's better than no distinction.

File: internal/controllers/virtual_devices_controller.go [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+		return fiber.NewError(fiber.StatusUnauthorized, "error occurred when fetching user")
+	}
+
+	if user.EthereumAddress == nil {
+		return fiber.NewError(fiber.StatusUnauthorized, "User does not have an Ethereum address on file.")
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), uTokenID)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "failed to get integration")
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:114]: Translate again. Note that we're just using this to check that the integration _exists_.

File: internal/controllers/virtual_devices_controller.go [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid tokenID provided")
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), uTokenID)
+	if err != nil {
+		return helpers.GrpcErrorToFiber(err, "failed to get integration")
+	}
+
+	user, err := vc.usersClient.GetUser(c.Context(), &pb.GetUserRequest{
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:175]: Do we check that this person owns the NFT? That it exists?

I think we could consolidate the following logic:

1. User on file
2. User has address
3. Vehicle NFT exists
4. User owns NFT

I'm sure it's annoying to write twice.

Comment (elffjs) [reply internal/controllers/virtual_devices_controller.go:175]: Let's store `userAddr = common.HexToAddress(*user.EthereumAddress)`.

File: internal/controllers/virtual_devices_controller.go [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+	if err != nil {
+		return fiber.NewError(fiber.StatusInternalServerError, "Couldn't not verify signature.")
+	}
+
+	pubKey, err := crypto.UnmarshalPubkey(sigPublicKey)
+	if err != nil {
+		return fiber.NewError(fiber.StatusInternalServerError, "Couldn't not verify signature.")
+	}
+
+	payloadVerified := bytes.Equal(crypto.PubkeyToAddress(*pubKey).Bytes(), common.HexToAddress(*user.EthereumAddress).Bytes())
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:205]: Then you can just check `crypto.PubkeyToAddress(*pubKey) = userAddr`.

File: internal/controllers/virtual_devices_controller.go (hunk 1 above)
Comment (elffjs) [reply internal/controllers/virtual_devices_controller.go:148]: 3. `POST virtual-device/mint/{integrationId}/{vehicleId}` (which order?) is also a possibility. The advantage here is that it's easy to search in Nginx, which is powerful.

Comment (0xdev22) [reply internal/controllers/virtual_devices_controller.go:148]: Good point, will make the change

Comment (elffjs) [reply internal/controllers/virtual_devices_controller.go]: I am iffy on this as a URL pattern.

File: internal/controllers/virtual_devices_controller.go [hunk 7]
```diff
@@ -0,0 +1,233 @@
...
+	"github.com/savsgio/gotils/bytes"
+	"github.com/volatiletech/null/v8"
+	"github.com/volatiletech/sqlboiler/v4/types"
+)
+
+type VirtualDeviceController struct {
+	Settings     *config.Settings
+	DBS          func() *db.ReaderWriter
+	log          *zerolog.Logger
+	integSvc     services.DeviceDefinitionIntegrationService
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:30]: Do we use this for anything?

File: internal/controllers/virtual_devices_controller.go [hunk 8]
```diff
@@ -0,0 +1,233 @@
...
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
+	}
+
+	pubRaw, err := crypto.UnmarshalPubkey(pub)
+	if err != nil {
+		vc.log.Err(err).Msg("Error occurred marshalling recovered public public key")
+		return fiber.NewError(fiber.StatusBadRequest, "Couldn't verify signature.")
+	}
+
+	payloadVerified := bytes.Equal(crypto.PubkeyToAddress(*pubRaw).Bytes(), userAddr.Bytes())
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:226]: These are both `common.Address`; you can do `crypto.PubkeyToAddress(*pubRaw) == userAddr`.

File: internal/controllers/virtual_devices_controller.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+	deviceDefSvc services.DeviceDefinitionService
+	usersClient  pb.UserServiceClient
+}
+
+type SignVirtualDeviceMintingPayloadRequest struct {
+	VehicleNode int `json:"vehicleNode"`
+	Credentials struct {
+		AuthorizationCode string `json:"authorizationCode"`
+	} `json:"credentials"`
+	OwnerSignature string `json:"ownerSignature"`
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:40]: Let's use `hexutil.Bytes`. That saves us the work of having to parse this out.

File: internal/controllers/virtual_devices_controller.go [hunk 10]
```diff
@@ -0,0 +1,233 @@
...
+}
+
+// GetVirtualDeviceMintingPayload godoc
+// @Description gets the payload for to mint virtual device given an integration token ID
+// @Tags        integrations
+// @Produce     json
+// @Param       integrationNode path int true "token ID"
+// @Param       vehicleID path int true "vehicle ID"
+// @Success     200 {array} signer.TypedData
+// @Router      /mint/:integrationNode/:vehicleID [get]
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:113]: I think we want the full path after `v1` here.

File: internal/controllers/virtual_devices_controller.go [hunk 11]
```diff
@@ -0,0 +1,233 @@
...
+	})
+	if err != nil {
+		vc.log.Debug().Err(err).Msg("error occurred when fetching user")
+		return helpers.GrpcErrorToFiber(err, "error occurred when fetching user")
+	}
+
+	vid, err := strconv.ParseInt(vehicleNode, 10, 64)
+	if err != nil {
+		return fiber.NewError(fiber.StatusBadRequest, "invalid vehicleNode provided")
+	}
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:135]: Can we consolidate the parsing?

File: internal/controllers/virtual_devices_controller.go [hunk 12]
```diff
@@ -0,0 +1,233 @@
...
+
+	if err = vc.verifyUserAddressAndNFTExist(c.Context(), user, vid, rawIntegrationNode); err != nil {
+		return err
+	}
+
+	integration, err := vc.deviceDefSvc.GetIntegrationByTokenID(c.Context(), integrationNode)
+	log.Println("avbbbbbbbbb", integration)
```
Comment (elffjs) [inline internal/controllers/virtual_devices_controller.go:142]: I am going to remove these.

File: internal/controllers/virtual_devices_controller.go (hunk 9 above)
Comment (elffjs) [reply internal/controllers/virtual_devices_controller.go]: Eh it's too strict, I forgot that it requires `0x` in the front.

File: internal/controllers/virtual_devices_controller.go (hunk 7 above)
Comment (elffjs) [reply internal/controllers/virtual_devices_controller.go]: Deleted it.

**Pull request merged**